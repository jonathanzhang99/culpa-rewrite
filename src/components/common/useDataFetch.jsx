import { useContext, useEffect, useReducer } from "react";

import { AuthContext } from "components/common/Authentication";

function dataFetchReducer(state, action) {
  switch (action.type) {
    case "FETCH_START":
      return {
        ...state,
        isError: false,
        isLoading: true,
      };
    case "FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case "FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true,
        error: action.payload,
      };
    default:
      throw new Error();
  }
}

/**
 * This hook is the main way to fetch api data from the backend. The local
 * state will store all retrieved data. If the parent AuthContext holds a
 * userToken, it will automatically be passed along and automatically removed
 * if the server throws an error on the token.
 *
 * @param {string} url - the endpoint from where to fetch data
 * @param {Object} initialData - default data to display while loading
 */
export default function useDataFetch(url, initialData) {
  const [state, dispatch] = useReducer(dataFetchReducer, {
    data: initialData,
    isError: false,
    isLoading: false,
  });
  const { authToken, isLoggedIn, logout } = useContext(AuthContext);

  useEffect(() => {
    const getHeaders = () => {
      const headers = {
        Accept: "application/json",
      };

      if (isLoggedIn) {
        headers.Authorization = `Bearer ${authToken}`;
      }

      return headers;
    };

    const fetchData = async () => {
      dispatch({ type: "FETCH_START" });

      const headers = getHeaders();
      const response = await fetch(url, {
        method: "GET",
        headers,
      });

      try {
        const results = await response.json();

        if (!response.ok) {
          const authErrorCodes = [401, 403, 422];
          if (authErrorCodes.includes(response.status)) {
            logout(results.error);
          }

          dispatch({ type: "FETCH_FAILURE", payload: results.error });
        } else {
          dispatch({ type: "FETCH_SUCCESS", payload: results });
        }
      } catch (err) {
        dispatch({ type: "FETCH_FAILURE", payload: err });
      }
    };

    fetchData();
  }, [authToken, isLoggedIn, logout, url]);

  return state;
}
