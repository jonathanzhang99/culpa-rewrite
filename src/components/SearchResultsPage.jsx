import PropTypes from "prop-types";
import queryString from "query-string";
import React from "react";
import { useLocation } from "react-router-dom";
import { Divider, Grid, Header } from "semantic-ui-react";

import { CourseDisplayLink } from "components/common/CourseDisplay";
import { DepartmentDisplayLink } from "components/common/DepartmentDisplay";
import ErrorComponent from "components/common/ErrorComponent";
import LoadingComponent from "components/common/LoadingComponent";
import { ProfessorDisplayLink } from "components/common/ProfessorDisplay";
import useDataFetch from "components/common/useDataFetch";

const propTypesSearchResults = {
  results: PropTypes.arrayOf(
    PropTypes.shape({
      childKey: PropTypes.string.isRequired,
      departments: PropTypes.arrayOf(
        PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired,
        })
      ).isRequired,
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      type: PropTypes.oneOf(["professor", "course"]).isRequired,
    })
  ).isRequired,
};

function SearchResults({ results }) {
  if (!results.length) return <p> No result found </p>;
  return (
    <Grid>
      {results.map(({ childKey, departments, id, title, type }) => (
        <Grid.Row key={childKey}>
          <Grid.Column width={6}>
            {type === "professor" ? (
              <ProfessorDisplayLink fullName={title} professorId={id} />
            ) : (
              <CourseDisplayLink courseId={id} courseName={title} />
            )}
          </Grid.Column>
          <Grid.Column width={10}>
            {departments.map(({ id: departmentId, name }, index) => (
              <span key={`department-${departmentId}`}>
                <DepartmentDisplayLink
                  departmentId={departmentId}
                  departmentName={name}
                />
                {index !== departments.length - 1 ? "," : ""}
              </span>
            ))}
          </Grid.Column>
        </Grid.Row>
      ))}
    </Grid>
  );
}

export default function SearchResultsPage() {
  const { search } = useLocation();

  const {
    data: { professorResults, courseResults },
    isLoading,
    isError,
  } = useDataFetch(`/api/search${search}`, {
    professorResults: [],
    courseResults: [],
  });

  if (isLoading || isError) {
    return isLoading ? <LoadingComponent /> : <ErrorComponent />;
  }

  const { query } = queryString.parse(search);

  return (
    <>
      <Header>
        Showing search results for <em> {query} </em>
      </Header>
      <Divider />
      <Header> Professors </Header>
      <SearchResults results={professorResults} />
      <Header> Courses </Header>
      <SearchResults results={courseResults} />
    </>
  );
}

SearchResults.propTypes = propTypesSearchResults;
