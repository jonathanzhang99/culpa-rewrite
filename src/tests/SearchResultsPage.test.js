import { render } from "@testing-library/react";
import React from "react";
import { MemoryRouter } from "react-router-dom";

import useDataFetch from "components/common/useDataFetch";
import SearchResultsPage from "components/SearchResultsPage";

jest.mock("../components/common/useDataFetch");

afterEach(() => {
  useDataFetch.mockReset();
});

describe("SearchResultsPage Components", () => {
  const testCases = [
    {
      name: "both professors and courses found",
      testProfessorResults: [
        {
          childKey: "professor-2339",
          departments: [
            {
              id: 29,
              name: "Computer Science",
            },
          ],
          last: "false",
          id: 2339,
          title: "Nakul Verma",
          type: "professor",
        },
      ],
      testCourseResults: [
        {
          childKey: "course-1",
          departments: [
            {
              id: 6,
              name: "Computer Science",
            },
          ],
          id: 1,
          title: "User Interface Design",
          type: "course",
        },
      ],
    },
    {
      name: "only professors found",
      testProfessorResults: [
        {
          childKey: "professor-2339",
          departments: [
            {
              id: 29,
              name: "Computer Science",
            },
          ],
          last: "false",
          id: 2339,
          title: "Nakul Verma",
          type: "professor",
        },
      ],
      testCourseResults: [],
    },
    {
      name: "only courses found",
      testProfessorResults: [],
      testCourseResults: [
        {
          childKey: "course-1",
          departments: [
            {
              id: 6,
              name: "Computer Science",
            },
          ],
          id: 1,
          title: "User Interface Design",
          type: "course",
        },
      ],
    },
    {
      name: "no results found",
      testProfessorResults: [],
      testCourseResults: [],
    },
  ];

  testCases.forEach(({ name, testProfessorResults, testCourseResults }) => {
    test(name, () => {
      useDataFetch.mockImplementation(() => ({
        data: {
          professorResults: testProfessorResults,
          courseResults: testCourseResults,
        },
        isError: false,
        isLoading: false,
      }));
      const snapshot = render(
        <MemoryRouter>
          <SearchResultsPage />
        </MemoryRouter>
      );
      expect(snapshot).toMatchSnapshot();
      expect(useDataFetch).toHaveBeenCalledTimes(1);
    });
  });
});
