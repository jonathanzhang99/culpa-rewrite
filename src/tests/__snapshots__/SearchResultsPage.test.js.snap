// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResultsPage Components both professors and courses found 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
        Showing search results for 
        <em>
           
           
        </em>
      </div>
      <div
        class="ui divider"
      />
      <div
        class="ui header"
      >
         Professors 
      </div>
      <div
        class="ui grid"
      >
        <div
          class="row"
        >
          <div
            class="six wide column"
          >
            <a
              href="/professor/2339"
            >
              <span>
                Nakul Verma
              </span>
            </a>
            <img
              class="ui avatar left spaced image"
              src="gold-nugget.svg"
            />
            <img
              class="ui avatar left spaced image"
              src="silver-nugget.svg"
            />
          </div>
          <div
            class="ten wide column"
          >
            <span>
              <a
                href="/department/29"
              >
                <span>
                   
                  Computer Science
                   
                </span>
              </a>
              
            </span>
          </div>
        </div>
      </div>
      <div
        class="ui header"
      >
         Courses 
      </div>
      <div
        class="ui grid"
      >
        <div
          class="row"
        >
          <div
            class="six wide column"
          >
            <a
              href="/course/1"
            >
              <span>
                
                 
                User Interface Design
              </span>
            </a>
          </div>
          <div
            class="ten wide column"
          >
            <span>
              <a
                href="/department/6"
              >
                <span>
                   
                  Computer Science
                   
                </span>
              </a>
              
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
      Showing search results for 
      <em>
         
         
      </em>
    </div>
    <div
      class="ui divider"
    />
    <div
      class="ui header"
    >
       Professors 
    </div>
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="six wide column"
        >
          <a
            href="/professor/2339"
          >
            <span>
              Nakul Verma
            </span>
          </a>
          <img
            class="ui avatar left spaced image"
            src="gold-nugget.svg"
          />
          <img
            class="ui avatar left spaced image"
            src="silver-nugget.svg"
          />
        </div>
        <div
          class="ten wide column"
        >
          <span>
            <a
              href="/department/29"
            >
              <span>
                 
                Computer Science
                 
              </span>
            </a>
            
          </span>
        </div>
      </div>
    </div>
    <div
      class="ui header"
    >
       Courses 
    </div>
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="six wide column"
        >
          <a
            href="/course/1"
          >
            <span>
              
               
              User Interface Design
            </span>
          </a>
        </div>
        <div
          class="ten wide column"
        >
          <span>
            <a
              href="/department/6"
            >
              <span>
                 
                Computer Science
                 
              </span>
            </a>
            
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchResultsPage Components no results found 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
        Showing search results for 
        <em>
           
           
        </em>
      </div>
      <div
        class="ui divider"
      />
      <div
        class="ui header"
      >
         Professors 
      </div>
      <p>
         No result found 
      </p>
      <div
        class="ui header"
      >
         Courses 
      </div>
      <p>
         No result found 
      </p>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
      Showing search results for 
      <em>
         
         
      </em>
    </div>
    <div
      class="ui divider"
    />
    <div
      class="ui header"
    >
       Professors 
    </div>
    <p>
       No result found 
    </p>
    <div
      class="ui header"
    >
       Courses 
    </div>
    <p>
       No result found 
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchResultsPage Components only courses found 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
        Showing search results for 
        <em>
           
           
        </em>
      </div>
      <div
        class="ui divider"
      />
      <div
        class="ui header"
      >
         Professors 
      </div>
      <p>
         No result found 
      </p>
      <div
        class="ui header"
      >
         Courses 
      </div>
      <div
        class="ui grid"
      >
        <div
          class="row"
        >
          <div
            class="six wide column"
          >
            <a
              href="/course/1"
            >
              <span>
                
                 
                User Interface Design
              </span>
            </a>
          </div>
          <div
            class="ten wide column"
          >
            <span>
              <a
                href="/department/6"
              >
                <span>
                   
                  Computer Science
                   
                </span>
              </a>
              
            </span>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
      Showing search results for 
      <em>
         
         
      </em>
    </div>
    <div
      class="ui divider"
    />
    <div
      class="ui header"
    >
       Professors 
    </div>
    <p>
       No result found 
    </p>
    <div
      class="ui header"
    >
       Courses 
    </div>
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="six wide column"
        >
          <a
            href="/course/1"
          >
            <span>
              
               
              User Interface Design
            </span>
          </a>
        </div>
        <div
          class="ten wide column"
        >
          <span>
            <a
              href="/department/6"
            >
              <span>
                 
                Computer Science
                 
              </span>
            </a>
            
          </span>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchResultsPage Components only professors found 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="ui header"
      >
        Showing search results for 
        <em>
           
           
        </em>
      </div>
      <div
        class="ui divider"
      />
      <div
        class="ui header"
      >
         Professors 
      </div>
      <div
        class="ui grid"
      >
        <div
          class="row"
        >
          <div
            class="six wide column"
          >
            <a
              href="/professor/2339"
            >
              <span>
                Nakul Verma
              </span>
            </a>
            <img
              class="ui avatar left spaced image"
              src="gold-nugget.svg"
            />
            <img
              class="ui avatar left spaced image"
              src="silver-nugget.svg"
            />
          </div>
          <div
            class="ten wide column"
          >
            <span>
              <a
                href="/department/29"
              >
                <span>
                   
                  Computer Science
                   
                </span>
              </a>
              
            </span>
          </div>
        </div>
      </div>
      <div
        class="ui header"
      >
         Courses 
      </div>
      <p>
         No result found 
      </p>
    </div>
  </body>,
  "container": <div>
    <div
      class="ui header"
    >
      Showing search results for 
      <em>
         
         
      </em>
    </div>
    <div
      class="ui divider"
    />
    <div
      class="ui header"
    >
       Professors 
    </div>
    <div
      class="ui grid"
    >
      <div
        class="row"
      >
        <div
          class="six wide column"
        >
          <a
            href="/professor/2339"
          >
            <span>
              Nakul Verma
            </span>
          </a>
          <img
            class="ui avatar left spaced image"
            src="gold-nugget.svg"
          />
          <img
            class="ui avatar left spaced image"
            src="silver-nugget.svg"
          />
        </div>
        <div
          class="ten wide column"
        >
          <span>
            <a
              href="/department/29"
            >
              <span>
                 
                Computer Science
                 
              </span>
            </a>
            
          </span>
        </div>
      </div>
    </div>
    <div
      class="ui header"
    >
       Courses 
    </div>
    <p>
       No result found 
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
